modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala

```diff
@@ -1172,11 +1173,38 @@ private[codegen] class SmithyToIR(model: Model, namespace: String) {
       s"Unhandled trait binding:\ntype: $tpe\nvalue: ${Node.printJson(node)}"
   }
 
-  private def unfoldTrait(tr: Trait): Hint.Native = {
+  private def unfoldTrait(shape: Shape, tr: Trait): Hint.Native = {
```

Comment:
I'd rather not modify this method. Instead, I'd create a new `Hint` type holding just the ShapeId and the Node (say `Hint.NativeDynamic` for the sake of the convo), and have a new method do `if isRecursive then createNativeDynamic(trait) else unfoldTrait(trait)` and call that where `unfoldTrait` is currently called.

Reply:
this is so much nicer, thanks!